src/peergos/server/space/JdbcUsageStore.java

```diff
@@ -0,0 +1,370 @@
+            insert.setBoolean(3, false);
+            insert.executeUpdate();
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private long getSpaceUsed(int userId) {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT total_bytes FROM userusage WHERE user_id = ?;")) {
+            insert.setInt(1, userId);
+            ResultSet resultSet = insert.executeQuery();
+            resultSet.next();
+            return resultSet.getLong(1);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public void confirmUsage(String username, PublicKeyHash writer, long usageDelta) {
+        int userId = getUserId(username);
+        long currentUsage = getSpaceUsed(userId);
+        try (PreparedStatement insert = conn.prepareStatement("UPDATE userusage SET total_bytes = ?, errored = ? " +
+                "WHERE user_id = ? AND total_bytes = ?;")) {
+            insert.setLong(1, currentUsage + usageDelta);
+            insert.setBoolean(2, false);
+            insert.setInt(3, userId);
+            insert.setLong(4, currentUsage);
+            int count = insert.executeUpdate();
+            if (count != 1)
+                throw new IllegalStateException("Didn't update one record!");
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+        clearPendingUsage(username, writer);
+    }
+
+    @Override
+    public void clearPendingUsage(String username, PublicKeyHash writer) {
+        int writerId = getWriterId(writer);
+        try (PreparedStatement insert = conn.prepareStatement("UPDATE pendingusage SET pending_bytes = ? " +
+                "WHERE writer_id = ?;")) {
+            insert.setLong(1, 0);
+            insert.setInt(2, writerId);
+            int count = insert.executeUpdate();
+            if (count != 1)
+                throw new IllegalStateException("Didn't update one record!");
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private int getPending(int writerId) {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT pending_bytes FROM pendingusage WHERE writer_id = ?;")) {
+            insert.setInt(1, writerId);
+            ResultSet resultSet = insert.executeQuery();
+            resultSet.next();
+            return resultSet.getInt(1);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public void addPendingUsage(String username, PublicKeyHash writer, int size) {
```

Comment:
I think the 2 sql statments can be combined into 1 command

Reply:
yes please

Reply:
like this:
INSERT INTO pendingusage (user_id, writer_id, pending_bytes) values (1, 2, 10)
ON CONFLICT (writer_id) DO UPDATE SET pending_bytes = pendingusage.pending_bytes + 1;